twap/main.go

```diff
@@ -0,0 +1,219 @@
+)
+
+type ArgsIterate struct {
+	TokenAddress string          `json:"token_address"`
+	ChainID      string          `json:"chain_id"`
+	NumSamples   int             `json:"num_samples"`
+	EAttest      json.RawMessage `json:"eAttest"`
+	TAttest      json.RawMessage `json:"tAttest"`
+	Whitelist    json.RawMessage `json:"whitelist"`
+}
+
+type ArgsTWAP struct {
+	EAttest   json.RawMessage `json:"eAttest"`
+	TAttest   json.RawMessage `json:"tAttest"`
+	Whitelist json.RawMessage `json:"whitelist"`
+}
+
+type Result struct {
+	Success bool
+	Error   string
+}
+
+type PriceSamples []price.Price
+
+type ResultPriceSamples struct {
+	Result       Result
+	PriceSamples PriceSamples
+}
+
+type ResultTWAP struct {
+	Result Result
+	TWAP   float64
+}
+
+func extractPriceSamples(
+	eAttest,
+	tAttest,
+	whitelist json.RawMessage,
+) (
+	PriceSamples,
+	error,
+) {
+	// bootstrap with empty samples if we don't have a transitive attestation
+	if tAttest == nil {
+		return PriceSamples{}, nil
+	}
+
+	verifyOut, err := as.VerifyAttestation(
+		as.HostVerifyAttestationInput{
+			EnclaveAttestedKey:    eAttest,
+			TransitiveAttestation: tAttest,
+			AcceptableMeasures:    whitelist,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify previous attestation: %w", err)
+	}
+
+	var fixedRep [][]byte
+	err = json.Unmarshal(verifyOut.RawClaims, &fixedRep)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal previous claims: %w", err)
+	}
+
+	prevResultData := fixedRep[3]
+	var prevResult ResultPriceSamples
+	err = json.Unmarshal(prevResultData, &prevResult)
+	switch {
+	case err != nil:
+		return nil, fmt.Errorf("could not unmarshal previous output: %w", err)
+	case !prevResult.Result.Success:
+		return nil, fmt.Errorf("previous run was an error: %w", err)
+	}
+
+	return prevResult.PriceSamples, nil
+
+}
+
+func getNewPriceSample(tokenAddress string, chainID string) (price.Price, error) {
+	req := as.HostHTTPRequestInput{
+		Method: "GET",
+		URL: fmt.Sprintf(
+			"https://app.steer.finance/api/token/price?tokenAddress=%s&chainId=%s",
+			tokenAddress,
+			chainID,
+		),
+	}
+	resp, err := as.HostFuncHTTPRequest(req)
+	if err != nil {
+		return price.Price{}, fmt.Errorf("making http request: %w", err)
+	}
+
+	steerData := struct {
+		Price float64 `json:"price"`
+	}{}
+
+	err = json.Unmarshal(resp.Body, &steerData)
+	if err != nil {
+		return price.Price{}, fmt.Errorf(
+			"unmarshaling coin gecko data: %w...%s",
+			err,
+			resp.Body,
+		)
+	}
+
+	now, err := as.TimeNow()
+	if err != nil {
+		return price.Price{}, fmt.Errorf("getting current time: %w", err)
+	}
+
+	return price.Price{
+		Value:     steerData.Price,
+		Timestamp: now,
+	}, nil
+}
+
+//export iteration
+func iteration(inputPtr, secretPtr uint64) uint64 {
+	var args ArgsIterate
+	inputData := as.Bytes(inputPtr)
+	err := json.Unmarshal(inputData, &args)
+	if err != nil {
+		outErr := fmt.Errorf("could not unmarshal args args: %w", err)
+		return emitErr(outErr.Error())
```

Comment:
FWIW, I am not a huge fan "emit" terminology.  It has a specific Prog Lang meaning (especially to those familiar with Node).  It made some sense in the first mode of computation that in the POC and before we came to some conclusion of the standard interface, but is kind of confusing when paired with the standard interface.  Specifically, you can "emit" things, but since we only sign and return the return value to the user, all the other stuff that was emitted goes to a black hole.

Reply:
OK makes sense. I'll update to `write` following the terminology in https://github.com/blocky/delphi/blob/8196fb486008cd32a623312a3e2a455eefbe0af5/test/live/testdata/kitchenSink-go/main.go#L268.